### PR TITLE
Add Enterprise Search Module

### PR DESCRIPTION
**Pull request by carlosdelest**
Introduces a new module for Enterprise Search. This module will be extended with additional Enterprise Search offerings for v8.8 and beyond.

To start, this module provides two CRUD APIs under the `_application` namespace. These offerings all require an enterprise or platinum license. To start they exist behind a feature flag (`xpack.ent-search.enabled`).

- Search Applications (`_application/search_application`), which will be a standard premium search experience going forward. 
- Behavioral Analytics (`_application/analytics`), which provides behavioral analytics easily integrated into indices and applications.

Some notes on design: 
- Search applications are stored in a system index, and we create an alias for the search application for use by search
- Analytics are powered by data streams without the need for an additional system index

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
I only really focused on java things & testing. I didn't review functionality.

**Review by carlosdelest — COMMENTED** [review]
Thanks for your review, @benwtrent ! 💯 

I've corrected some of the issues mentioned. Analytics related issues will be addressed in other commits.

Please keep the feedback coming!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java]: You're correct. In fact this was spotted in https://github.com/elastic/elasticsearch/pull/94752 - I'm getting the code from there and adding a YAML test to check this specific case. Thanks!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:111]: Addressed as part of https://github.com/elastic/elasticsearch/pull/94813.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:79]: Addressed as part of https://github.com/elastic/elasticsearch/pull/94813

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionRequestSerializingTests.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionRequestSerializingTests.java:24]: I made it not nullable in the class as part of https://github.com/elastic/elasticsearch/pull/94813.
Indeed, the RestController is sending an empty array is the param is not provided.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java
Comment (benwtrent) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java:93]: Don't use clone. I know its used elsewhere, but it is much preferred to use `Arrays.copyOf`.
```suggestion
        this.indices = Arrays.copyOf(indices, indices.length);
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java
Comment (benwtrent) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java:94]: Why are we sorting this? I cannot find anything that expects a specific order?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java
Comment (benwtrent) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: What happens if this is called, gets passed the `removeAlias` part, but before `deleteSearchApplication` is called, somebody calls `putSearchApplication`?

From what I can tell, it is possible to have the aliases stay around with the search application deleted.

The higher level transport calls are simply `HandledTransportActions` and even though you are using `clusterService` you could be seeing an "old" version of cluster state that is not fully committed.

I 100% expect this race condition to be hit. If we have a way out to clean it up, some warnings, etc. or if we are OK about it, please comment.

**Review by benwtrent — COMMENTED** [review]
some more comments :)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/TransportPutAnalyticsCollectionAction.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/TransportPutAnalyticsCollectionAction.java]: Updated. Thank for pointing this out.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: > In any case calling deleteSearchApplication and putSearchApplication externally on the same resource without synchronisation is already a race condition?

Not that we have detected for single user usage. The Kibana UI is not performing those operations on quick succession. 

Of course multiple users could potentially run into this issue as Ben described.

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: > Of course multiple users could potentially run into this issue as Ben described.

Right but then it's the same issue for index creation/deletion, mapping updates, ...

Comment (benwtrent) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: ML has ran into the same issue multiple times. The key things are:

 - Where it is REALLY important for synchronization, use a lock (rare)
 - Where things can get out of whack, make sure its recoverable via the user (deletion, recreation, warnings, etc.)
 - Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed. If creating the alias fails, try not to be in a bad state, so delete the search application and indicate that creating the alias failed.

I think things that are CRUD for things that cannot have tons of race conditions should be master node operations. 

My main concern is that we think about the failure paths and bad states we can easily get into once we are making two disjoint things rely on each other.

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: > Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed.

I like this approach. I think that's what we had first but we reordered to create/delete the alias first so that the action checks the permissions on the indices.
We could reorder and ensure that every failing paths during creation should ensure that the SearchApplication document is deleted in the system index? @carlosdelest wdyt?

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: It's good to settle on always doing the SearchApplication related action first. We can add to the code paths checks to ensure Search Application doc is deleted on failure as well.

In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: > In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

We could do that but I wonder if this is really needed. Could we fail the request and send back the error that only the search application was deleted?

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: Sounds good, will start working on it and create a PR with the changes 👍

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: @benwtrent would that alleviate your concerns here?

Comment (benwtrent) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: @jimczi I suppose. As long as the situation is recoverable and things are clear if it causes things to break in the future (like, the user creating a search application with the same name later). 

ML actually has a nightly cleanup script to clear out orphan objects. If this gets more common, something like that here might be required.

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: I've created [this draft](https://github.com/elastic/elasticsearch/pull/94924) with the changes. Looks mostly ok, but for the update case.

For updating, if we have already updated the Search App but the alias fails, then we should restore the original Search App. Which means we need to get it first. Which can go wrong.

The code is tricky already with this change (maybe that's my fault), but I think these error conditions are adding complexity to the code and there'll be multiple test cases to be done to ensure everything is working perfectly. And even so, we won't get rid of race conditions when users are interacting simultaneously.

Would another option be preferable to this? What are your thoughts?

**Review by cbuescher — COMMENTED** [review]
I didn't get through all the changes in this PR, and since I'm out for a couple of days I'm just adding the few questions and comments I have. None of this should block this PR, so I'm merely leaving these as suggestions or comments, nothing to block merging once somebody else approves.

File: docs/reference/search-application/apis/list-search-applications.asciidoc
Comment (ioanatia) [reply docs/reference/search-application/apis/list-search-applications.asciidoc:25]: This is very similar to the `q` parameter we have in the [search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-search-api-path-params), which in the docs is described as:

> q
(Optional, string) Query in the Lucene query string syntax.

The way we end up using the `q` parameter is by constructing a [Query string query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html) to query the system index storing the search applications. The supported syntax of the query string in the list Search Applications API should be the same to what we have for `q` parameter in the search API ([docs for query string syntax](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-syntax)), which is why they have similar documentation.

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json
Comment (afoucret) [reply x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json:47]: Hey @cbuescher,

The final mapping have change since we wrote this PR and this label object have been removed.
It will be merged as part of the implementation of the event intake API.
You can see a preview of the final mapping here: 

https://github.com/elastic/elasticsearch/blob/b589719df9fceed6f8755c985234533fbdefd633/x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json

File: docs/reference/search-application/apis/list-search-applications.asciidoc
Comment (afoucret) [reply docs/reference/search-application/apis/list-search-applications.asciidoc:47]: 💯 We should probably avoid `foo`ing the documentation!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionAction.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionAction.java:113]: This is used in `AnalyticsCollectionServiceTests` which is not in the same package.
Also it seems like that getting the collections is a legit public API for the list collection response.
WDYT?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollection.java
Comment (afoucret) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollection.java:80]: It feels like we will need to persist the Analytics Collection somewhere (probably into a system index) in the future to support new features (user profile resolutions, ...). Implementing the toXContent interface will likely help us to implement this storage.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:355]: After a discussion with @jimczi and @ioanatia, for this PR we'd like to:

- Keep the alias creation as is and document the race condition as a known issue for a beta release
- Create a followup issue to resolve this issue, required before the feature is GA. 

Please let us know if there is any other followup work required to address this at this time. Thank you!

File: docs/reference/search-application/apis/put-search-application.asciidoc
Comment (kderusso) [reply docs/reference/search-application/apis/put-search-application.asciidoc:27]: We decided to default to `false` because this call will also be used regularly to update existing search applications. Furthermore the behavior was consistent with `create` flags in many other calls such as the [Indices Put Template](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-template.html) call.

File: docs/reference/search-application/apis/list-search-applications.asciidoc
Comment (kderusso) [reply docs/reference/search-application/apis/list-search-applications.asciidoc:47]: Good catch, addressing this: https://github.com/elastic/elasticsearch/pull/94996

**Review by javanna — APPROVED** [review]

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

**Comment by afoucret** [comment]
🎉

**Pull request merged**
